### PR TITLE
Refactor for Namespace and Project

### LINKED ISSUE #1470: Cannot deploy kubernetes resources to a specific namespace when connected to default
### Description
It seems that is not possible to deploy namespaced kubernetes fragments when the kube context is connected to an other namespace than one specified in yaml files 
### Info

* f-m-p version : 4.0.0-M2
* Maven version (`mvn -v`) : 3.6.0
```

```
* Kubernetes / OpenShift setup and version :
* If it's a bug, how to reproduce : 

1. set current namespace in ~kube/config to default
2. run fabric8:resources to create a namespace, a deployment and a service in the created namespace 

* If it's a feature request, what is your use case :
* Sample project : *[GitHub Clone URL]*

### LINKED ISSUE #1608: namespace.yml fragment doesn't get converted to project kind

### LINKED ISSUE #1611: Allow creating namespaces from xml config
Currently there's no option to create a namespace via xml config option.

### PR DESCRIPTION
**Pull request by dev-gaur**
Fixes #1608
Fixes #1611
Fixes #1470

### CODE REVIEW

**Comment by dev-gaur** [comment]
What this PR does :

If a user provides a resource fragment for namespace, add <namespace>NAME</namespace> in the xml-config in pom.xml or provide a -Dfabric8.namespace=NAME parameter for ResourceMojo , 

a namespace resource would be generated. And all the other component resource's metadata will have a "namespace : NAME " entry.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+package io.fabric8.maven.enricher.standard.openshift;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.kubernetes.api.model.Namespace;
+import io.fabric8.kubernetes.api.model.NamespaceSpec;
+import io.fabric8.maven.core.config.PlatformMode;
+import io.fabric8.maven.enricher.api.BaseEnricher;
+import io.fabric8.maven.enricher.api.MavenEnricherContext;
+import io.fabric8.openshift.api.model.*;
```
Comment (rohanKanojia) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: avoid wildcard imports

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            ProjectStatus status = new ProjectStatusBuilder()
+                    .withPhase(namespace.getStatus().getPhase()).build();
+
+            namespace.getStatus().getAdditionalProperties()
+                    .forEach((k, v) -> status.setAdditionalProperty(k, v));
+        }
+
+        Project project = builder.build();
+
+        namespace.getAdditionalProperties()
```
Comment (rohanKanojia) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: Could you please tell why is this required?

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    /**
+     * This method will annotate all the items in the KubernetesListBuilder with the
+     * created new namespace or project.
+     * @param platformMode
+     * @param builder
+     */
+    @Override
+    public void enrich(PlatformMode platformMode, KubernetesListBuilder builder) {
+
+        builder.accept(new TypedVisitor<DeploymentBuilder>() {
```
Comment (rohanKanojia) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java]: We're only editing metadata, I wonder if we can just iterate over ObjectMetaBuilder rather than iterating over all types individually.

File: core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java [hunk 4]
```diff
@@ -247,11 +247,14 @@ public Builder withRemotes(List<String> remotes) {
             return this;
         }
 
+        public Builder withNameSpace(String s) {
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java]: Nope. It is deliberately withNamespace(String s) , for cases where we need to alter the namespace name from the default config.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            return;
+        }
+
+        if (!KubernetesResourceUtil.checkForKind(builder, NAMESPACE_KINDS)) {
+            String type = getConfig(Config.type);
+            if ("project".equalsIgnoreCase(type) || "namespace".equalsIgnoreCase(type)) {
+                if (platformMode == PlatformMode.kubernetes) {
+                    try {
+                        createKubernetesClient().namespaces().list().getItems().forEach(namespace -> {
+                            if(namespace.getMetadata().getName().equals(name)) {
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java]: fair point

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                if(item instanceof Namespace) {
+                    Project project = convert(item);
+                    removeItemFromKubernetesBuilder(builder, item);
+                    builder.addToProjectItems(project);
+                }
+            }
+        }
+    }
+
+    private Project convert(HasMetadata item) {
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: aah, good catch!

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java (hunk 2 above)
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: If the namespace resource object has any additional properties, these should also be passed on to the project resource being created. 

You can refer to Project and Namespace resource classes in kubernetes-model

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java (hunk 3 above)
Comment (dev-gaur) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java]: good catch, thanks !

I tried with a visitor on HasMetaData first, that didn't work. I think it works on builder classes only

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java [hunk 7]
```diff
@@ -52,17 +52,7 @@
...
-import io.fabric8.openshift.api.model.DoneableImageStream;
-import io.fabric8.openshift.api.model.ImageStream;
-import io.fabric8.openshift.api.model.ImageStreamSpec;
-import io.fabric8.openshift.api.model.OAuthClient;
-import io.fabric8.openshift.api.model.Project;
-import io.fabric8.openshift.api.model.ProjectRequest;
-import io.fabric8.openshift.api.model.Route;
-import io.fabric8.openshift.api.model.TagReference;
-import io.fabric8.openshift.api.model.Template;
+import io.fabric8.openshift.api.model.*;
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: Avoid star import

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java [hunk 8]
```diff
@@ -132,9 +122,11 @@ private void applyEntity(Object dto, String sourceName) throws Exception {
             applyReplicationController((ReplicationController) dto, sourceName);
         } else if (dto instanceof Service) {
             applyService((Service) dto, sourceName);
-        } else if (dto instanceof Namespace) {
+/*        } else if (dto instanceof Namespace) {
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: If it is not necessary just remove it.

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java [hunk 9]
```diff
@@ -972,6 +964,20 @@ public boolean applyNamespace(Namespace entity) {
         return false;
     }
 
+    /**
+     * Creates and return a project in openshift
+     * @param project
+     * @return
+     */
+    public boolean applyProject(Project project) {
+/*        getOpenShiftClient().projects().
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: If it is not necessary just remove it

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 10]
```diff
@@ -459,22 +458,23 @@ private void lateInit() {
         if (runtimeMode.equals(RuntimeMode.openshift)) {
             Properties properties = project.getProperties();
             if (!properties.contains(DOCKER_IMAGE_USER)) {
-                String namespace = clusterAccess.getNamespace();
+                String namespace = this.namespace!= null && this.namespace.length() > 0 ?
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Add space between `namespace` and operator `!=`

Comment (lordofthejars) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: I recommend using `!namespace.isEmpty()` rather than check size

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java (hunk 7 above)
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: done

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java (hunk 8 above)
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: Aah. I missed it. Removed.

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java (hunk 10 above)
Comment (dev-gaur) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Done

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java (hunk 9 above)
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: aah I'm sorry I forgot to omit this part

File: core/src/main/java/io/fabric8/maven/core/service/ApplyService.java [hunk 11]
```diff
@@ -52,25 +44,20 @@
...
-import static io.fabric8.maven.core.util.kubernetes.KubernetesHelper.getOrCreateMetadata;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static io.fabric8.maven.core.util.kubernetes.KubernetesHelper.*;
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/ApplyService.java]: Another star import.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+     * @param builder
+     */
+    @Override
+    public void enrich(PlatformMode platformMode, KubernetesListBuilder builder) {
+
+        builder.accept(new TypedVisitor<ObjectMetaBuilder>() {
+            @Override
+            public void visit(ObjectMetaBuilder builder) {
+                String name = config.getNamespace();
+                if (name == null || name.length() == 0) {
```
Comment (lordofthejars) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java]: use `isEmpty` instead of length

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.maven.core.config.PlatformMode;
+import io.fabric8.maven.core.config.ResourceConfig;
+import io.fabric8.maven.core.handler.HandlerHub;
+import io.fabric8.maven.core.util.Configs;
+import io.fabric8.maven.core.util.kubernetes.KubernetesResourceUtil;
+import io.fabric8.maven.enricher.api.BaseEnricher;
+import io.fabric8.maven.enricher.api.MavenEnricherContext;
+import io.fabric8.openshift.api.model.*;
```
Comment (lordofthejars) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/DefaultNamespaceEnricher.java]: Star import

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+
+        if (namespace.getSpec() != null) {
+            NamespaceSpec namespaceSpec = namespace.getSpec();
+            ProjectSpec projectSpec = new ProjectSpec();
+            if (namespaceSpec.getFinalizers() != null) {
+                projectSpec.setFinalizers(namespaceSpec.getFinalizers());
+            }
+            namespaceSpec.getAdditionalProperties()
+                    .forEach((k, v) -> projectSpec.setAdditionalProperty(k, v));
```
Comment (lordofthejars) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: Give a meaning name in the vars

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+            builder.withSpec(projectSpec);
+        }
+
+        if (namespace.getStatus() != null) {
+            ProjectStatus status = new ProjectStatusBuilder()
+                    .withPhase(namespace.getStatus().getPhase()).build();
+
+            namespace.getStatus().getAdditionalProperties()
+                    .forEach((k, v) -> status.setAdditionalProperty(k, v));
```
Comment (lordofthejars) [inline enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: Meaning name to the vars

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 16]
```diff
@@ -20,22 +20,9 @@
...
 import io.fabric8.maven.core.access.ClusterAccess;
 import io.fabric8.maven.core.access.ClusterConfiguration;
-import io.fabric8.maven.core.config.MappingConfig;
-import io.fabric8.maven.core.config.OpenShiftBuildStrategy;
-import io.fabric8.maven.core.config.PlatformMode;
-import io.fabric8.maven.core.config.ProcessorConfig;
-import io.fabric8.maven.core.config.Profile;
-import io.fabric8.maven.core.config.ResourceConfig;
-import io.fabric8.maven.core.config.RuntimeMode;
+import io.fabric8.maven.core.config.*;
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Star import

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 17]
```diff
@@ -20,22 +20,9 @@
...
 import io.fabric8.maven.core.handler.HandlerHub;
-import io.fabric8.maven.core.model.GroupArtifactVersion;
-import io.fabric8.maven.core.util.FileUtil;
-import io.fabric8.maven.core.util.MavenUtil;
-import io.fabric8.maven.core.util.ProfileUtil;
-import io.fabric8.maven.core.util.ResourceClassifier;
-import io.fabric8.maven.core.util.ResourceFileType;
-import io.fabric8.maven.core.util.ResourceUtil;
-import io.fabric8.maven.core.util.ValidationUtil;
+import io.fabric8.maven.core.util.*;
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Star import

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 18]
```diff
@@ -53,29 +40,20 @@
...
 import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.MojoFailureException;
-import org.apache.maven.plugins.annotations.Component;
-import org.apache.maven.plugins.annotations.LifecyclePhase;
-import org.apache.maven.plugins.annotations.Mojo;
-import org.apache.maven.plugins.annotations.Parameter;
-import org.apache.maven.plugins.annotations.ResolutionScope;
+import org.apache.maven.plugins.annotations.*;
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Star import

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 19]
```diff
@@ -53,29 +40,20 @@
...
 import org.apache.maven.project.MavenProjectHelper;
 import org.apache.maven.shared.filtering.MavenFileFilter;
 import org.apache.maven.shared.filtering.MavenFilteringException;
 
+import javax.validation.ConstraintViolationException;
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Star import

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 20]
```diff
@@ -459,22 +437,23 @@ private void lateInit() {
...
         }
-        handlerHub = new HandlerHub(
-            new GroupArtifactVersion(project.getGroupId(), project.getArtifactId(), project.getVersion()),
-            project.getProperties());
     }
 
     private KubernetesList generateResources(PlatformMode platformMode, List<ImageConfiguration> images, File remoteResources)
         throws IOException, MojoExecutionException {
 
+        if (namespace != null && namespace.length() > 0) {
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: use `isEmpty()`

**Comment by dev-gaur** [comment]
sorry I made a blunder with the star imports! I've changed the settings in my IDE now. Shouldn't be a problem in future.

File: core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java]: I'm not able to understand what you're saying. I'm just asking to rename this method. I see it as a typo.

File: enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java (hunk 2 above)
Comment (rohanKanojia) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: Umm,  still not able to understand why we need this. I don't see any additional properties field here:

https://kubernetes.io/docs/tasks/administer-cluster/namespaces-walkthrough/#create-new-namespaces

As far as I know, these additionalProperties are used internally to store additional information. Or they are set if some portions of object were not able to get parsed properly.

Comment (rohanKanojia) [reply enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/openshift/ProjectEnricher.java]: I don't think we should care about porting them also during conversion.

File: core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/config/ResourceConfig.java]: Aah, I understood something else.
